packages/go_router/lib/src/route.dart

```diff
@@ -346,6 +344,14 @@ abstract class ShellRouteBase extends RouteBase {
   /// Returns the key for the [Navigator] that is to be used for the specified
   /// immediate sub-route of this shell route.
   GlobalKey<NavigatorState> navigatorKeyForSubRoute(RouteBase subRoute);
+
+  /// Creates or updates an optional state object to be associated with an
+  /// active shell route.
+  ///
+  /// The existing state can be accessed via [ShellRouteContext.shellRouteState].
+  Object? updateShellState(BuildContext context, GoRouterState state,
```

Comment:
I am not entirely sure about this API. It looks like the return value was feed directly in buildPage or buildWidget. why not just put the logic in `buildPage` directly? It looks like there is some caching in RouteBuilder._shellRouteState. Can that be put into the StatefulNavigationShell instead?

Reply:
If it has to be this way (which I doubt), consider declaring a generic Type `T` for ShellRouteBase so that you don't need to use `Object?`